tfmigrate/test_xmv_action_test.go

```diff
@@ -0,0 +1,165 @@
+			nrWildcards: 0,
+		},
+		{
+			desc:        "Simple wildcardChar for a resource",
+			resource:    NewStateXMvAction("aws_security_group.*", "aws_security_group.$1"),
+			nrWildcards: 1,
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.desc, func(t *testing.T) {
+			got := tc.resource.nrOfWildcards()
+			if got != tc.nrWildcards {
+				t.Errorf("Number of wildcards for %d is not expected %s", got, tc.resource)
+			}
+		})
+	}
+}
+
+func TestSingleResourceStateMoves(t *testing.T) {
+	cases := []struct {
+		desc            string
+		stateList       []string
+		inputXMvAction  *StateXMvAction
+		outputMvActions []*StateMvAction
+	}{
+		{
+			desc:      "Simple resource no wildcardChar",
+			stateList: nil,
+			inputXMvAction: &StateXMvAction{
+				source:      "aws_security_group.foo",
+				destination: "aws_security_group.foo2",
+			},
+			outputMvActions: []*StateMvAction{
+				{
+					source:      "aws_security_group.foo",
+					destination: "aws_security_group.foo2",
+				},
+			},
+		},
+		{
+			desc:      "Simple resource with wildcardChar",
+			stateList: []string{"aws_security_group.foo"},
+			inputXMvAction: &StateXMvAction{
+				source:      "aws_security_group.*",
+				destination: "module.security_group[\"$1\"].sg",
+			},
+			outputMvActions: []*StateMvAction{
+				{
+					source:      "aws_security_group.foo",
+					destination: "module.security_group[\"foo\"].sg",
+				},
+			},
+		},
+		{
+			desc:      "Simple module name refactor with wildcardChar",
+			stateList: []string{"module.security_group[\"foo\"].sg"},
+			inputXMvAction: &StateXMvAction{
+				source:      "module.security_group[\"*\"].sg",
+				destination: "module.sg[\"$1\"].sg",
+			},
+			outputMvActions: []*StateMvAction{
+				{
+					source:      "module.security_group[\"foo\"].sg",
+					destination: "module.sg[\"foo\"].sg",
+				},
+			},
+		},
+		{
+			desc:      "No matching resources in state",
+			stateList: []string{"aws_vpc.foo"},
+			inputXMvAction: &StateXMvAction{
+				source:      "aws_security_group.*",
+				destination: "module.security_group[\"$1\"].sg",
+			},
+			outputMvActions: []*StateMvAction{},
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.desc, func(t *testing.T) {
+			got, err := tc.inputXMvAction.getStateMvActionsForStateList(tc.stateList)
+			// Should be location agnostic
+			if err != nil {
+				t.Errorf("Encountered error %v", err)
+			}
+			if len(got) != len(tc.outputMvActions) {
+				t.Errorf("Did not get same amount of matches: %d, %d", len(got), len(tc.outputMvActions))
+			}
+
+			for i, outputMvAction := range got {
+				if tc.outputMvActions[i].source != outputMvAction.source {
+					t.Errorf("Mistmatching output sources %d: %v, %v", i, tc.outputMvActions[i].source, outputMvAction.source)
+				}
+				if tc.outputMvActions[i].destination != outputMvAction.destination {
+					t.Errorf("Mistmatching output destination %d: %v, %v", i, tc.outputMvActions[i].destination, outputMvAction.destination)
+				}
+			}
```

Comment:
The `reflect.DeepEqual()` is an easy way to compare slices.
https://pkg.go.dev/reflect#DeepEqual

Reply:
I believe I did it this way such that I could print out helpful error messages when unittest fail. It does make sense to use reflect.DeepEqual as it makes the test much more readable but in order to keep output clear I have added a `String` function with StateMvAction receiver but I'm unclear where to place it. I have placed it in `state_xmv_action_test.go`  eventhough it takes a struct from `state_mv_action.go`. Reason being that String() is a public function but I don't really implemented this with usage besides testing in mind and if someone wants to have a real string function they can remove this knowing that it is only used for tests currently.

Reply:
If your concern is only for pretty printing, since we have already depended on some 3rd-party libraries (go-spew and go-cmp), how about this?

```diff
diff --git a/tfmigrate/test_xmv_action_test.go b/tfmigrate/test_xmv_action_test.go
index eb4e7ae..085c5f2 100644
--- a/tfmigrate/test_xmv_action_test.go
+++ b/tfmigrate/test_xmv_action_test.go
@@ -2,10 +2,10 @@ package tfmigrate

 import (
        "context"
-       "fmt"
-       "reflect"
        "testing"

+       "github.com/davecgh/go-spew/spew"
+       "github.com/google/go-cmp/cmp"
        "github.com/minamijoyo/tfmigrate/tfexec"
 )

@@ -91,7 +91,7 @@ func TestGetStateMvActionsForStateList(t *testing.T) {
                        outputMvActions: []*StateMvAction{
                                {
                                        source:      "aws_security_group.foo",
-                                       destination: "aws_security_group.foo2",
+                                       destination: "aws_security_group.foo3",
                                },
                        },
                },
@@ -185,15 +185,9 @@ func TestGetStateMvActionsForStateList(t *testing.T) {
                                t.Errorf("Encountered error %v", err)
                        }

-                       if !reflect.DeepEqual(got, tc.outputMvActions) {
-                               t.Errorf("got: %v, expected: %v", got, tc.outputMvActions)
+                       if diff := cmp.Diff(got, tc.outputMvActions, cmp.AllowUnexported(StateMvAction{})); diff != "" {
+                               t.Errorf("got = %s, want = %s, diff = %s", spew.Sdump(got), spew.Sdump(tc.outputMvActions), diff)
                        }
-
                })
        }
 }
-
-// At this time purely for testing purposes if a real string function is needed for code this can be replaced.
-func (a StateMvAction) String() string {
-       return fmt.Sprintf("mv %s %s", a.source, a.destination)
-}
```

```
bash-5.1# go test -v ./tfmigrate -run=TestGetStateMvActionsForStateList
=== RUN   TestGetStateMvActionsForStateList/Simple_resource_no_wildcardChar
    test_xmv_action_test.go:189: got = ([]*tfmigrate.StateMvAction) (len=1 cap=1) {
         (*tfmigrate.StateMvAction)(0xc0000da5c0)({
          source: (string) (len=22) "aws_security_group.foo",
          destination: (string) (len=23) "aws_security_group.foo2"
         })
        }
        , want = ([]*tfmigrate.StateMvAction) (len=1 cap=1) {
         (*tfmigrate.StateMvAction)(0xc0000da420)({
          source: (string) (len=22) "aws_security_group.foo",
          destination: (string) (len=23) "aws_security_group.foo3"
         })
        }
        , diff =   []*tfmigrate.StateMvAction{
                &{
                        source:      "aws_security_group.foo",
        -               destination: "aws_security_group.foo2",
        +               destination: "aws_security_group.foo3",
                },
          }

```

Reply:
Cool, I like this approach.